cdap-proto/src/main/java/co/cask/cdap/proto/id/EntityId.java

```diff
@@ -152,21 +153,63 @@ public MetadataEntity toMetadataEntity() {
    * metadataEntity.
    */
   public static <T extends EntityId> T fromMetadataEntity(MetadataEntity metadataEntity) {
-    EntityType entityType = EntityType.valueOf(metadataEntity.getType().toUpperCase());
-    if ((entityType == EntityType.APPLICATION || entityType == EntityType.PROGRAM) &&
-      metadataEntity.getValue(MetadataEntity.VERSION) == null) {
-      // if the EntityType is application or program and a version is not specified in the MetadataEntity then add it
-      if (entityType == EntityType.APPLICATION) {
-        metadataEntity = metadataEntity.append(MetadataEntity.VERSION, ApplicationId.DEFAULT_VERSION);
-      } else {
-        metadataEntity = MetadataEntity.ofNamespace(metadataEntity.getValue(MetadataEntity.NAMESPACE))
-          .append(MetadataEntity.APPLICATION, metadataEntity.getValue(MetadataEntity.APPLICATION))
-          .append(MetadataEntity.VERSION, ApplicationId.DEFAULT_VERSION)
-          .append(MetadataEntity.TYPE, metadataEntity.getValue(MetadataEntity.TYPE))
-          .append(MetadataEntity.PROGRAM, metadataEntity.getValue(MetadataEntity.PROGRAM));
+    EntityType.valueOf(metadataEntity.getType().toUpperCase());
+    return getKnownParentEntityId(metadataEntity);
+  }
+
+  /**
+   * Creates a valid known CDAP entity which can be considered as the parent for the MetadataEntity by walking up the
+   * key-value hierarchy of the MetadataEntity till a known CDAP {@link EntityType} is found.
+   *
+   * @param metadataEntity whose parent entityId needs to be found
+   * @return {@link EntityId} of the given metadataEntity
+   */
+  public static <T extends EntityId> T getKnownParentEntityId(MetadataEntity metadataEntity) {
+    EntityType entityType = findParentType(metadataEntity);
+    List<String> values = new LinkedList<>();
+    List<MetadataEntity.KeyValue> till = metadataEntity.getTill(entityType.toString());
+    String version = metadataEntity.containsKey(MetadataEntity.VERSION) ?
+      metadataEntity.getValue(MetadataEntity.VERSION) : ApplicationId.DEFAULT_VERSION;
+    if (entityType == EntityType.APPLICATION) {
+      till.add(new MetadataEntity.KeyValue(MetadataEntity.VERSION, version));
+    }
+    if (entityType == EntityType.PROGRAM) {
```

Comment:
I don't understand what you are doing here?

Reply:
Added comments to code to explain